### PR TITLE
Added undo and clear button on drawing panel

### LINKED ISSUE #12979: Better button placement for handwriting feature
**Is your feature request related to a problem? Please describe.**

When writing kanji characters, there is no button to wipe the board clean. You must go back stroke by stroke to the beginning to rewrite your character from scratch. There should be a reset/wipe board feature. 

Additionally, when hitting the arrow button to go back a stroke, it is very easy to turn off the writing pad altogether or toggle the flag menu. If you want to go back several strokes and hit the back button several times consecutively, you risk turning the writing pad off by accident then hitting the back button. This makes you go back to the previous card and not back a stroke. I've had it happen where I'm trying to get to the beginning of my stroke order So I'm hitting the back arrow a few times consecutively, and instead I accidentally turn off the writing pad and send myself back two or three cards.

**Describe the solution you'd like**
I would like to see a small dedicated handwriting menu open with a back button for stroke order, an erase button to clear the board completely, and an on and off toggle switch. This small handwritten menu should be away from all the other review function buttons such as the flag button or go back button.

**Describe alternatives you've considered**

None 

**Additional context**

This would be useful not just for Japanese but also Chinese, Korean - any language where there are complex ideographic characters with a stroke order. It would also permit me to practice writing the kanji several times by writing it then completely erasing it rather than write it once then have to go back stroke by stroke.

### PR DESCRIPTION
**Pull request by Aayush2111**
## Purpose / Description
Added undo and clear button on drawing panel

## Fixes
Fixes #12979 

## Approach
By adding undo button we can undo our last stroke and clear button will clear all the screen

## How Has This Been Tested?

Tested on physical device- android version 11

WhatsApp Image 2022-12-18 at 17 20 57

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by pratyaksh1610 — CHANGES_REQUESTED** [review]
🚀 Congratulations on your first PR ! . Some changes before the initial pass. I think you should remove the portion which you have commented for smooth review process and run `./gradlew ktlintFormat` to format the code as per the standards.

**Comment by Aayush2111** [comment]
> 

done, thanks! :)

File: AnkiDroid/src/main/res/layout/reviewer_whiteboard_editor.xml [hunk 1]
```diff
@@ -6,46 +6,64 @@
...
+    <Button
+        android:id="@+id/pen_color_green"
+        android:background="@color/material_green_900"
+        style="@style/reviewer_whiteboard_editor_button_style"/>
+
+    <Button
+        android:id="@+id/pen_color_black"
+        android:background="@color/black"
+        style="@style/reviewer_whiteboard_editor_button_style"/>
+
```
Comment (pratyaksh1610) [inline AnkiDroid/src/main/res/layout/reviewer_whiteboard_editor.xml:64]: Please remove these unnecessary lines of code.

**Review by pratyaksh1610 — CHANGES_REQUESTED** [review]

**Comment by pratyaksh1610** [comment]
@Aayush2111 I think a sidebar would be a better option in my opinion . Thanks !

**Comment by Aayush2111** [comment]
> @Aayush2111 I think a sidebar would be a better option in my opinion . Thanks !

like permanent sidebar on the left side of the screen or at bottom?

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt [hunk 2]
```diff
@@ -85,6 +87,7 @@ class Whiteboard(activity: AnkiActivity, handleMultiTouch: Boolean, inverted: Bo
 
     override fun onDraw(canvas: Canvas) {
         super.onDraw(canvas)
+        // Draw the current state of the canvas
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt:90]: unnecessary

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt [hunk 3]
```diff
@@ -321,12 +324,17 @@ class Whiteboard(activity: AnkiActivity, handleMultiTouch: Boolean, inverted: Bo
 
     fun onClick(view: View) {
         when (view.id) {
+
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt:327]: unnecessary blank line

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt [hunk 4]
```diff
@@ -538,6 +550,7 @@ class Whiteboard(activity: AnkiActivity, handleMultiTouch: Boolean, inverted: Bo
     init {
         val whitePenColorButton = activity.findViewById<Button>(R.id.pen_color_white)
         val blackPenColorButton = activity.findViewById<Button>(R.id.pen_color_black)
+
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt:553]: unnecessary blank line

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt [hunk 5]
```diff
@@ -42,6 +42,8 @@ import com.ichi2.utils.KotlinCleanup
 import com.mrudultora.colorpicker.ColorPickerPopUp
 import timber.log.Timber
 import java.io.FileNotFoundException
+import java.util.*
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.kt:45]: do we need this import?

File: AnkiDroid/src/main/res/layout/reviewer_whiteboard_editor.xml [hunk 6]
```diff
@@ -12,40 +12,53 @@
         style="@style/reviewer_whiteboard_editor_button_style"/>
 
     <Button
-        android:id="@+id/pen_color_black"
-        android:background="@color/black"
+        android:id="@+id/pen_color_yellow"
+        android:background="@color/material_yellow_500"
         style="@style/reviewer_whiteboard_editor_button_style"/>
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/res/layout/reviewer_whiteboard_editor.xml:17]: Why are you adding a new color and changing the colors order?

As you haven't described your approach on the commits or the PR description, I can't tell

File: AnkiDroid/src/main/res/layout/reviewer_whiteboard_editor.xml [hunk 7]
```diff
@@ -12,40 +12,53 @@
...
 
     <Button
         android:id="@+id/pen_color_red"
-        android:background="@color/material_red_500"
+        android:background="@color/flag_red"
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/res/layout/reviewer_whiteboard_editor.xml:21]: Same here. Why are you changing this to `flag_red`?

**Review by BrayanDSO — CHANGES_REQUESTED** [review]
Tested on SDK 33 emulator, Pixel 2

Functionality-wise, the main issues I see are:
1. the undo and clear canvas buttons don't appear on light themes,
    - image
2. The user can't scroll to the green and black colors you added
2. The color buttons close the whiteboard editor while the new buttons don't. I believe that it would be better to be consistent and not close the editor after tapping a color button.
3. *optional*: if you want to, making the undo button faded if there isn't anything to undo would be nice
4. *optional*: it would be good to be able to undo a `Clear canvas` action

About the PR organization itself:
1. There are some merge commits, which normally isn't a problem, but on AnkiDroid we strive to keep a linear git history, so we only do squash and rebase merges. This means that if you want your git history to be merged, you'd better rebase the PR. Otherwise, you can leave the merge commits as is, and the PR will be squashed when it is merged.
2. The commits could be more helpful
    - They should describe what the changes intend to do
    - They shouldn't have temporary changes as commenting a whole block of code just to test things
3.

**Comment by Aayush2111** [comment]
Working on the requested changes!

**Comment by Aayush2111** [comment]
> 1. the undo and clear canvas buttons don't appear on light themes,
-Resolved

> 2. The user can't scroll to the green and black colors you added
-Resolved

> 3. The color buttons close the whiteboard editor while the new buttons don't. I believe that it would be better to be consistent and not close the editor after tapping a color button.
- I guess it'd be better if clicking on colors close the editor as it will provide more space for us to draw while other buttons can stay as it is since it will be frustrating if we have to undo more than one stroke and manually opening editor again and again. I didn't change this as of now, but do let me know if you think editor shouldn't get close after tapping a color button, I will fix it.

> 4. _optional_: if you want to, making the undo button faded if there isn't anything to undo would be nice
-I have tried adding new drawable _ic_undo_selector.xml_ which will link to _ic_undo_disabled.xml_ (if android:state_enabled="false") and _ic_undo_white.xml_, but this doesn't work for some reason only _ic_undo_white.xml_ is working. I am not quite sure what I am missing extra?

**Comment by BrayanDSO** [comment]
> The user can't scroll to the green and black colors you added
Resolved

Here I still cannot do that

**Comment by Aayush2111** [comment]
> > The user can't scroll to the green and black colors you added
> > Resolved
> 
> Here I still cannot do that

I have removed the green color on platelet by default, working fine for me

screenshot: 
WhatsApp Image 2023-01-06 at 23 13 13
WhatsApp Image 2023-01-06 at 23 13 13

**Comment by BrayanDSO** [comment]
> I have removed the green color on platelet by default, working fine for me

Getting rid of things without need isn't a good way of solving problems IMO.

**Comment by Aayush2111** [comment]
> IMO.

I see, but other way would be to change the platelet structure completely as keeping all default colors with current platelet structure with extra two buttons now doesn't seem ideal to me.

Should I restructure the platelet then?

**Comment by BrayanDSO** [comment]
> Should I restructure the platelet then?

I think that this is a good idea

**Comment by Aayush2111** [comment]
> > Should I restructure the platelet then?
> 
> I think that this is a good idea



> > Should I restructure the platelet then?
> 
> I think that this is a good idea

WhatsApp Image 2023-01-07 at 00 42 33

Is something like this ideal? 
PS: sorry for bad drawing :P

**Comment by david-allison** [comment]
@Aayush2111 If you're around, I'm back and happy to work to get this to completion

**Comment by SanjaySargam** [comment]
Hey @Aayush2111 If you are not working on it then let me know. I would love to continue this implementation

**Comment by Aayush2111** [comment]
> Hey @Aayush2111 If you are not working on it then let me know. I would love to continue this implementation

Sure, go for it.

**Comment by SanjaySargam** [comment]
@Aayush2111 Thanks for informing

**Pull request closed without merging**